### PR TITLE
Custom protocol URI scheme to open the configurator (btfltcp://)

### LINKED ISSUE #2281: Button to open Betaflight
Hey Guys! Thanks for your help in the past.

I was wondering if anyone knows how to open betaflight using either a link (href) or a javascript function.

My goal is to open the app either on windows or mac and populate the navbar with "Manual Selection" with the port: (ie 'tcp://blah.blah')

Possible or no?
I dont really know where to ask and Google is not my friend today...

### PR DESCRIPTION
**Pull request by krasiyan**
Fixes #2281 

This PR makes it possible to open the Betaflight Configurator and immediately initialize a TCP connection via a custom URI protocol scheme: `btfltcp://`. The IP/URL (and optionally the port) of the TCP host (e.g. an ExpressLRS receiver connected via WiFi) can be passed in as: `btfltcp://{{address}}:{{port}}` - thus e.g. `btfltcp://10.0.0.1`, `btfltcp://10.0.0.1:5761` or even `btfltcp://elrs_rx.local`.

The scheme is automatically registered with the OS by the Configurator on all platforms during installation (currently Linux, Windows, OSX and Android). On all NW.js builds (Linux, Windows and OSX), the invocations with the scheme are handled both on a "cold start" as well as while the Configurator is running. On the Cordova (Android) builds, the [`cordova-plugin-customurlscheme`](https://www.npmjs.com/package/cordova-plugin-customurlscheme) NPM dep. is used to both register and handle the invocations with the custom URI scheme.

## Testing
The changes (as of commit 57c1883e3ea698d92f16770a5d7d36de53c8cce1) have been tested on the following platforms: 
- [x] Linux (deb)
- [x] Linux (rpm) 
- [x] Android (emulated)
- [x] Android (physical)
- [ ] OSX
- [x] Windows 

:information_source: Please note that in order to test this feature you must select the `enableURIScheme` experimental configuration option (it is _not_ necessary to restart the configurator afterwards):

btfltcp-option

## Next steps

- (question to maintainers :question: ) Should we do anything special for the SpeedyBee app or would they decide to handle the change on their own? It would be good to see if multiple Android applications can handle the same deep link / custom URI scheme (so that the user can choose which one to open).
- The ExpressLRS Web UI can be updated with a link to open the Betaflight Configurator (as simple as e.g. `<a href="btfltcp://{{ip}}">Open Betaflight Configurator</a>`)

## Demos

- Linux (cold start)

btfltcp-linux-cold

- Linux (warm start)
btfltcp-linux-warm

- Android (cold start)
btfltcp-android-cold

- Android (warm start)
btfltcp-android-warm

- Experimental configuration option toggling
btfltcp-experimental-option

### CODE REVIEW

File: cordova/yarn.lock [hunk 1]
```diff
@@ -669,6 +669,11 @@ cordova-plugin-chrome-apps-storage@^1.0.4:
   resolved "https://registry.yarnpkg.com/cordova-plugin-chrome-apps-storage/-/cordova-plugin-chrome-apps-storage-1.0.4.tgz#f83b1e16e54b774ba375f6ff5378fda829b161fd"
   integrity sha1-+DseFuVLd0ujdfb/U3j9qCmxYf0=
 
+cordova-plugin-customurlscheme@^5.0.2:
```
Comment (krasiyan) [inline cordova/yarn.lock:672]: Reviewer note - I noticed that Cordova was adding this automatically when running `cordova plugin add ...` locally - so I assumed it was missed out in the past and decided to commit it as well.

File: src/js/protocol_uri.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    if (GUI.isCordova()) {
+        handleCordovaOpenEvents();
+        return;
+    }
+
+    const command = `"${process.execPath}" $_URL_`;
+
+    console.log(`Registering ${protocolScheme} to '${command}'`);
+    try {
+        const ProtocolRegistry = require('protocol-registry');
```
Comment (krasiyan) [inline src/js/protocol_uri.js:20]: TODO: This dynamic `require` seems to prevent `rollup` from including `protocol-registry` in the build; as a result the NW.js builds don't properly register the schema.

File: cordova/yarn.lock [hunk 3]
```diff
@@ -714,6 +719,11 @@ cordova-plugin-whitelist@^1.3.4:
   resolved "https://registry.yarnpkg.com/cordova-plugin-whitelist/-/cordova-plugin-whitelist-1.3.4.tgz#31938545c7c3e7de35c20ab08c2c3afa06e8a3f9"
   integrity sha512-EYC5eQFVkoYXq39l7tYKE6lEjHJ04mvTmKXxGL7quHLdFPfJMNzru/UYpn92AOfpl3PQaZmou78C7EgmFOwFQQ==
 
+cordova-plugin-zeroconf@^1.4.2:
```
Comment (krasiyan) [inline cordova/yarn.lock:722]: Reviewer note - I noticed that Cordova was adding this automatically when running `cordova plugin add ...` locally - so I assumed it was missed out in the past and decided to commit it as well.

File: src/js/protocol_uri.js (hunk 2 above)
Comment (krasiyan) [reply src/js/protocol_uri.js]: Final update: :relieved: I dropped `protocol-registry` in favor of implementing everything in our `gulpfile`. :) This is a much cleaner setup overall - the only slight drawback being that one has to install/re-install the configurator for the URI schema to register. Thus during local development the URI schema will resolve to the _installed_ configurator and not to the current development build. Of course (if needed) one can manually update their `.deskop`/`.plist`/registry to point to the development executable.

**Comment by jennypavlova** [comment]
✅  Checked on M1 Max: OSX Sonoma 14.3.1

File: src/js/uri_scheme.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    $('#port').val('manual').trigger('change');
+    $('#port-override').val(tcpUrl).trigger('change');
+    $('div.connect_controls a.connect').trigger('click');
+}
+
+
+function handleNWJSArgv() {
+    const connectionString = nw?.App?.argv?.[0] || '';
```
Comment (chmelevskij) [inline src/js/uri_scheme.js:57]: in what case is this passed in used? Can you provide example usage?

Just wondering if it won't clash with something like `--inspect` oar other debug flags for example.

File: src/js/uri_scheme.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    if (connectionString) {
+        connectTcp(connectionString, 'cold start');
+    }
+}
+
+function handleNWJSOpenEvents() {
+    nw?.App?.on('open', (connectionString) => {
+        connectTcp(connectionString, 'NW.js "open" event');
+    });
```
Comment (chmelevskij) [inline src/js/uri_scheme.js:67]: When is this fired exactly?

File: src/js/uri_scheme.js (hunk 4 above)
Comment (krasiyan) [reply src/js/uri_scheme.js:57]: @chmelevskij `nw.App.argv` will be present every time the configurator is _cold_ started with any CLI argument. Thus cold starts through the `btfltcp://` URI scheme end up triggering invocations like `./betaflight-configurator btfltcp://ip` and  `nw.App.argv` gets set to `['btfltcp://...']`. [Here are the NW.js docs on `argv`](https://docs.nwjs.io/en/latest/References/App/#appargv) (it seems some of the NW.js-specific arguments such as `--nwapp` are already filtered out by default).

However you are absolutely right that we might end up running into edge-cases if the user modifies their default execution command (in the `.desktop`/plist/registry). A manual invocation (on Linux) with e.g. `/opt/betaflight/betaflight-configurator/betaflight-configurator --disable-gpu --someTestArgument btltcp://10.0.0.1` results in a `nw.App.argv` value of `['--disable-gpu', '--someTestArgument', 'btltcp://10.0.0.1']` (thus the newly added functionality breaks; but there is no impact for NW.js). 

I think the easiest solution might be to always get the last element in the array instead of the first one - thus:

```suggestion
    const connectionString = (nw?.App?.argv? || []).at(-1);
```
(since the URI scheme always seems to be added at the end of the array)

I will test that across the platforms and update the MR - thank you! :)

File: src/js/uri_scheme.js (hunk 5 above)
Comment (krasiyan) [reply src/js/uri_scheme.js:67]: @chmelevskij The `open` event is fired when the URI scheme is invoked while the configurator is already running. [As per the NW.js docs](https://docs.nwjs.io/en/latest/References/Command%20Line%20Options/#about-command-line-arguments):

> If there is already one running instance of your app, the full command line of the second instance will be passed to the existing instance by sending open event to the App object.

Coming back [to your earlier comment](https://github.com/betaflight/betaflight-configurator/pull/3831#discussion_r1525475582) - it seems that the event handler is always called with a single string which contains the executable path along with all CLI arguments - e.g. if I run `betaflight-configurator --someTestArg btfltcp://10.0.0.2` while the configurator is running I see a `connectionString` value similar to:

```
/opt/betaflight/betaflight-configurator/betaflight-configurator --someTestArg --disable-features=nw2 --user-data-dir=/home/USER/.config/betaflight-configurator --no-sandbox --no-zygote --flag-switches-begin --flag-switches-end --nwapp=/opt/betaflight/betaflight-configurator btfltcp://10.0.0.2
```
(the behavior is indeed a bit inconsistent in regards to `nw.App.argv` and apparently there is no filtering of the NW.js arguments - so that's why I went with for regex which extract just the `btfltcp://{ip}` segment)

File: src/js/uri_scheme.js (hunk 4 above)
Comment (krasiyan) [reply src/js/uri_scheme.js]: Re-tested and updated the MR - the aforementioned change seems to work as expected on all platforms.

Comment (chmelevskij) [reply src/js/uri_scheme.js]: Can we make it named argument then instead? That would de-couple it from any issues all together. Something like `--bf-uri` or similar? we already do something similar for the build parts https://github.com/betaflight/betaflight-configurator/blob/master/gulpfile.js#L146

File: src/js/uri_scheme.js (hunk 5 above)
Comment (chmelevskij) [reply src/js/uri_scheme.js:67]: Ah, fair, ignore my previous comment. If we have something to just check that url is correct seems fine to me 😊

**Review by chmelevskij — APPROVED** [review]

File: src/js/uri_scheme.js (hunk 4 above)
Comment (krasiyan) [reply src/js/uri_scheme.js]: @chmelevskij Unfortunately from what I checked, NW.js won't let us name the argument (I guess they intended it to be more or less hidden - as it is just appended to the end of the execution string). 

---

However indeed we could add a separate dedicated `--tcp-url={{IP/URL}}` argument for anyone who might want to launch the configurator from the CLI. Should we maybe open a separate issue for expanding its CLI interface? That way we can use the opportunity to also add e.g. a help menu and a few more arguments. I can imagine that this might be useful for power users/manufacturers. For example some rudimentary functionality for backup management might be nice - such as: 
- `./betaflight-configurator --tcp-url=... --diff-all > backup.txt` (establish the connection and pipe the `diff all` CLI output)
- `cat backup.txt | ./betaflight-configurator --tcp-url=... --load-diff` (execute the input from a pipe in the CLI)

Implementation-wise - NW.js also allows us to execute a script in a "pure" Node.js context via [the `node-main` script script in the manifest](https://docs.nwjs.io/en/latest/References/Manifest%20Format/#node-main). There is actually a very basic implementation of `node-main` in [a past revision of this MR](https://github.com/betaflight/betaflight-configurator/compare/897a64f9f744a9e8fc409573b0aac32e5e4fed6f..13deeb1075cefaf09f07b76ae26a70681fcdb9f9) (I tried it because of an external dependency for the URI schema management which I later discarded).

File: assets/windows/installer.iss [hunk 6]
```diff
@@ -167,4 +169,10 @@ begin
...
     end;
-end;
\ No newline at end of file
+end;
+
+[Registry]
+Root: HKA; Subkey: "Software\Classes\{#UriScheme}"; ValueType: "string"; ValueData: "URL:Custom Protocol"; Flags: uninsdeletekey
+Root: HKA; Subkey: "Software\Classes\{#UriScheme}"; ValueType: "string"; ValueName: "URL Protocol"; ValueData: ""
+Root: HKA; Subkey: "Software\Classes\{#UriScheme}\DefaultIcon"; ValueType: "string"; ValueData: "{app}\{cm:AppName}.exe,0"
+Root: HKA; Subkey: "Software\Classes\{#UriScheme}\shell\open\command"; ValueType: "string"; ValueData: """{app}\{cm:AppName}.exe"" ""%1"""
```
Comment (McGiverGim) [inline assets/windows/installer.iss:178]: I don't know if it's clearer to put the UriScheme directly into the file (the same for Linux, Android, etc.).

We did as constant in the Blackbox for file extensions for example:
https://github.com/betaflight/blackbox-log-viewer/blob/8419b1b815d74602c8cf92eb0439aacda1d960d8/assets/windows/installer.iss#L37-L43

Until now we only injected some values in the file for Android because they depending on the version. This seems more a "fixed" thing.

I see it reads it from the package.json. Is a standard tag of package.json? I can't find it in the documentation.
It's easier to read if we put it directly in the different files, but if we decide to change the value we will need to do a "search", but I suppose this will not happen. I don't know what think others. Opinions?

**Review by McGiverGim — COMMENTED** [review]
It seems good to me, only one doubt about harcoding or not the uriScheme.

File: assets/windows/installer.iss (hunk 6 above)
Comment (krasiyan) [reply assets/windows/installer.iss:178]: I definitely agree that the dynamic injection of the URI scheme might have been overengineered. The scheme will likely never change - especially if/once it is used in e.g. the ELRS web UI. I will update the MR unless anyone has any objections.

As for the `package.json` - indeed the newly added `uriScheme` is a custom/non-standardized field.

**Comment by haslinghuis** [comment]
image

should not be there when connected.

File: assets/windows/installer.iss (hunk 6 above)
Comment (krasiyan) [reply assets/windows/installer.iss]: PR updated - we no longer do dynamic substitutions for the `uriScheme`.

**Review by McGiverGim — APPROVED** [review]

**Comment by krasiyan** [comment]
> image
> 
> should not be there when connected.

Thanks for noticing @haslinghuis ! Indeed we had a bug where the `#port-override-option` was not being properly hidden upon connection establishment (I have since fixed it so it's both hidden on connection and shown after the disconnect callback). I also updated the screen recordings to show the new UI behavior.

File: src/js/serial_backend.js [hunk 7]
```diff
@@ -237,6 +238,10 @@ function finishClose(finishedCallback) {
     // unlock port select & baud
     $('div#port-picker #port').prop('disabled', false);
     if (!GUI.auto_connect) $('div#port-picker #baud').prop('disabled', false);
+    const selectedPort = $('div#port-picker #port option:selected');
+    if (selectedPort?.data()?.isManual) {
+        $('#port-override-option').show();
+    }
```
Comment (haslinghuis) [inline src/js/serial_backend.js:244]: Should we call `GUI.updateManualPortVisibility()` instead?

**Comment by haslinghuis** [comment]
Requires testing [on OSX].

**Comment by Pancronos** [comment]
Trying to test this pr but after installing a portable version here on windows 11 https://github.com/betaflight/betaflight-configurator/suites/22069555096/artifacts/1360150995 there is no way thru elrs or directly on BF configurator (manual selection) to use btfltcp://...
A more detailed use for portable windows version is appreciated.

**Comment by McGiverGim** [comment]
It's the installer who associates the uri to the Configurator. Maybe we can look for a workaround for portables.

**Comment by Pancronos** [comment]
> It's the installer who associates the uri to the Configurator. Maybe we can look for a workaround for portables.

I was 99% sure, but i love portables, may author can set a script to install or un-install the missing steps. Thanks.

**Comment by Pancronos** [comment]
Installed the windows installer version to test, typing btfltcp://10.0.0.1 request to open BF. All ok. 
All tab seems to work except Ports and VTX tab hang on waiting for data 
image
I also have some concern on the flow of operations. Example i power my FC (a speedybee stack) via usb that also provide power the elrs, If configurator has set auto-connect what happens? Will i run 2 configurator one via usb/com and one via elrs? Which has priority/preference? Is this PR intented to be "on the field" or for portable usage ? Again the idea is great, but i think needs some more in depth usage consideration.

**Review by haslinghuis — CHANGES_REQUESTED** [review]
Usability issues needs to be addressed. We are already in RC phase, so want to mark this as experimental feature as it takes time for real public testing to take place and discover other edge cases I would suggest to implement this as an option like we have done for mDNS.

File: src/js/serial_backend.js (hunk 7 above)
Comment (krasiyan) [reply src/js/serial_backend.js]: Definitely makes sense - thank you for the suggestion! I also took the liberty to update the `GUI.updateManualPortVisibility` logic to only show the `#port-override-option` (manual port input) if a connection has not been already established (based on `isConnected`).

File: src/js/serial_backend.js [hunk 8]
```diff
@@ -90,6 +90,7 @@ export function initializeSerialBackend() {
         let portName;
         if (selectedPort.data().isManual) {
             portName = $('#port-override').val();
+            $('#port-override-option').hide();
```
[the remark this answers is not in the document]
Comment (krasiyan) [reply src/js/serial_backend.js]: Updated similarly to https://github.com/betaflight/betaflight-configurator/pull/3831#discussion_r1546958509 the above suggestion. Also added an extra `GUI.updateManualPortVisibility` invocation in the `abortConnection` function to properly visualize the manual port input on an aborted/unsuccessful connection.

**Comment by krasiyan** [comment]
> Usability issues needs to be addressed. We are already in RC phase, so want to mark this as experimental feature as it takes time for real public testing to take place and discover other edge cases I would suggest to implement this as an option like we have done for mDNS.

@haslinghuis That is indeed a great idea - thank you! :) I have since added the new `enableURIScheme` configuration optional (disabled by default):

btfltcp-experimental-option

**Comment by krasiyan** [comment]
@Pancronos 

> > It's the installer who associates the uri to the Configurator. Maybe we can look for a workaround for portables.
> 
> I was 99% sure, but i love portables, may author can set a script to install or/and un-install the missing steps. 

Indeed - the feature at hand would not work for the portable executables since the installer is the one associating the URI scheme with the configurator binary installation path (on the OS level). 

I think it is _theoretically_ possible to implement that feature for the portable binaries - using e.g. something like the [`protocol-registry`](https://www.npmjs.com/package/protocol-registry) NPM dependency to register the URI scheme after the configurator has booted up (or once the configuration is selected).
(In fact I [initially implemented it that way](https://github.com/betaflight/betaflight-configurator/compare/897a64f9f744a9e8fc409573b0aac32e5e4fed6f..13deeb1075cefaf09f07b76ae26a70681fcdb9f9) but dropped the approach due to [some of the issues I ran into](https://github.com/betaflight/betaflight-configurator/pull/3831#discussion_r1522338479)).

However IMHO there are some drawbacks to doing permanent URI scheme associations with a portable executable:
- The executable can easily be moved / removed - which would leave a "dangling" URI scheme association in the OS.
- Some users might be wary of a portable executable doing "permanent" changes to their OS (such as creating `.desktop`/`.plist` files on Linux/OSX and editing the OS registry on Windows).
  - Some of the issues I ran into with [`protocol-registry`](https://www.npmjs.com/package/protocol-registry) are also related to this point. Normally the NW.js app runs in a sandboxed environment which doesn't let it do certain broader changes in the OS. I'm worried that the only potential workaround might be to run the configurator as root/admin which is _definitely_ a bad practice. 
- We might have to maintain more code/logic as the installer and runtime implementation would be quite different. 

That being said - I'm open to explore the possibilities as long as the maintainer don't think any of the above drawbacks might be a deal breaker. 

> Thanks. Btw it looks a cool feature with a future for development and integration with other software. Compliment to the author.

Thank you! :blush:

**Comment by krasiyan** [comment]
@Pancronos Thank you for testing and evaluating the feature! :pray: 

> Installed the windows installer version to test, typing btfltcp://10.0.0.1 request to open BF. All ok. All tab seems to work except Ports and VTX tab hang on waiting for data

Would you mind sharing your Betaflight (firmware) and ExpressLRS versions? Also does the issue occur only when connecting via the URI scheme or also on `master` (/ the latest releases) when manually connecting to `10.0.0.1`?

On my side I couldn't consistently reproduce it for the tabs you mentioned but I suppose there _might_ be something as I occasionally saw some MSP timeouts in the development console when browsing through the configurator.

btfltcp-ports-configuration-vtx

If this bug is also present on master / the latest releases I can suggest we open a new issue since there might be a problem with the underlaying connection mechanism (which wasn't really changed in this MR). 

> I also have some concern on the flow of operations. Example i power my FC (a speedybee stack) via usb that also provide power the elrs, If configurator has set auto-connect what happens? 

That is indeed a very good question! From what I have noticed, on a _cold start_ the URI scheme would take precedence over the "auto connect" feature (since it seems to be faster to start establishing the network connection than to discover all USB devices). However that might not always be the case - thus what I can propose is that we automatically disable the "auto-connect" upon a URI scheme invocation?

> Will i run 2 configurator one via usb/com and one via elrs? 

AFAIK that's not really possible since there is only one configurator instance at any given time (be it a "cold start" or a "warm start"). As such the configurator can connect to either a TCP address or a USB device. 

>  Is this PR intented to be "on the field" or for portable usage ?  Again the idea is great, but i think needs some more in depth usage consideration.

To be honest the sole use-case I considered is the "portable" one (e.g. via the receiver's WiFi network with a shortcut in the ExpressLRS web UI). I wouldn't see much value in doing a TCP-based connection if the FC is already plugged to the host via USB (since I assume the USB connection to be more stable + versatile).
However if you can think of any other potential use-cases that might need an adjustment in the implementation I would be very happy to take a look :relaxed:

File: src/js/serial_backend.js [hunk 9]
```diff
@@ -148,9 +149,10 @@ export function initializeSerialBackend() {
 
                     function onFinishCallback() {
                         finishClose(toggleStatus);
+                        GUI.updateManualPortVisibility();
                     }
 
-                    mspHelper.setArmingEnabled(true, false, onFinishCallback);
+                    mspHelper?.setArmingEnabled(true, false, onFinishCallback);
```
Comment (haslinghuis) [inline src/js/serial_backend.js:155]: Race condition?

**Review by haslinghuis — APPROVED** [review]
Love this level of professional contributions :dagger:

File: src/js/serial_backend.js (hunk 9 above)
Comment (krasiyan) [reply src/js/serial_backend.js:155]: Indeed - the steps to reproduce it are:

1. Start a new instance of the configurator (it must not have been connected to an FC yet as otherwise `mspHelper` will be defined)
    - I suppose the same can be reproduced with a slow/unresponsive FC
3. Connect to a TCP address that won't respond (or will be very slow in doing so) - e.g. in the context of these changes - `10.0.0.2` when in the WiFi network of an ELRS RX
4. Click connect and then subsequently click disconnect

Here is a small demo:

setArmingEnabled-race-condition

**Comment by Pancronos** [comment]
image
://github.com/betaflight/betaflight-configurator/assets/6061804/2ab870a2-92f4-4cc6-8ad2-15b892d5d6a9)
image
Don't know why but know after connecting ELRS (my drone is with pc-usb that powers elrs) doesn't work with 10.0.0.1 but it works with elrs .local i attached console log where is goes in timeout on some tabs

**Comment by Pancronos** [comment]
Just tried btfltcp://10.0.0.2  request to open BFConfigurator but it return port error (no configurator windows were open)
image

**Comment by Pancronos** [comment]
Tried to use only wifi + usb (no ethernet) seems return error connecting to BF server, then i tried to remove usb from the FC and it's a no go
[ajdoecjhplhoahggfmgflfnlechcljoc-1712252264147.log](https://github.com/betaflight/betaflight-configurator/files/14875222/ajdoecjhplhoahggfmgflfnlechcljoc-1712252264147.log)
[ajdoecjhplhoahggfmgflfnlechcljoc-1712252141275.log](https://github.com/betaflight/betaflight-configurator/files/14875223/ajdoecjhplhoahggfmgflfnlechcljoc-1712252141275.log)
[ajdoecjhplhoahggfmgflfnlechcljoc-1712252010447.log](https://github.com/betaflight/betaflight-configurator/files/14875224/ajdoecjhplhoahggfmgflfnlechcljoc-1712252010447.log)

**Comment by Pancronos** [comment]
@krasiyan Please let me know how i can help testing, I really like your idea, but i see some conflict, I think the issue i saw are due to multiple connections to same object. 
If you create a more proprietary connection, like some static ip that use same class of elrs like 10.0.0.>1 would be much better, moreover a specific local dns like BTFLT.local. This should solve issues with connections, Second i would implement/force in configurator menu the option to use BTFLTCP://
Basically leave to the user the option (unless i type btfltcp://xx.x.xx.xxx in browser) to start a specific connection, trying to select how to connect.
Mainly the question is when i should use BTFLTCP:// over any other exsisting connections? A check if any other connection com-port, usb, tcp or whatever is needed otherwise conflict will be there.

**Comment by haslinghuis** [comment]
@krasiyan

- closing this as it did not make it in time.
- as we are moving to PWA we can use URL directly.

Thanks for your contribution as the code quality was professional.

**Comment by krasiyan** [comment]
Absolutely no problem @haslinghuis (and sorry for the slow progress in the final stages here). Thank you and the others for the collaboration! :) 

Looking forward to the PWA version :rocket:
